### PR TITLE
🔥 Remove force convert for the response body

### LINKED ISSUE #1703: Dio throws wrong error type
### Package

dio

### Version

5.0.1

### Output of `flutter doctor -v`

```shell
❯ flutter doctor -v
[✓] Flutter (Channel stable, 3.7.5, on macOS 13.0.1 22A400 darwin-x64, locale en-AU)
    • Flutter version 3.7.5 on channel stable at /Users/timnew/Workspace/flutter_sdks/production
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision c07f788888 (6 days ago), 2023-02-22 17:52:33 -0600
    • Engine revision 0f359063c4
    • Dart version 2.19.2
    • DevTools version 2.20.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/timnew/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14B47b
    • CocoaPods version 1.12.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] VS Code (version 1.75.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.58.0

[✓] Connected device (3 available)
    • iPhone 14 Pro (mobile) • 8B1B6C6A-211B-431E-9BD3-7EEBF6D33029 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-16-1 (simulator)
    • macOS (desktop)        • macos                                • darwin-x64     • macOS 13.0.1 22A400 darwin-x64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 105.0.5195.125

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

### Dart Version

2.19.2

### Steps to Reproduce

1. Calling API, which is expected to return JSON when succeeded
2. API returns 403 due to token issue, which `content-type` as `text/html` instead of `application/json` (Due to WAF or so)
3. Expected to have `DioError[badResponse]` but `got DioError[unknown]`

### Expected Result

Dio throws `DioError[badResponse]`

### Actual Result

Dio throws `DioErro[unkown]`
```
flutter: Error: DioError [unknown]: null
Error: FormatException: Unexpected character (at character 1)
error code: 1020
^
flutter: Stack:
#0      _ChunkedJsonParser.fail (dart:convert-patch/convert_patch.dart:1383:5)
#1      _ChunkedJsonParser.parseNumber (dart:convert-patch/convert_patch.dart:1250:9)
#2      _ChunkedJsonParser.parse (dart:convert-patch/convert_patch.dart:915:22)
#3      _parseJson (dart:convert-patch/convert_patch.dart:35:10)
#4      JsonDecoder.convert (dart:convert/json.dart:610:36)
#5      JsonCodec.decode (dart:convert/json.dart:216:41)
#6      jsonDecode (dart:convert/json.dart:155:10)
#7      _decodeJson
#8      SyncTransformer.transformResponse
sync_transformer.dart:123
<asynchronous suspension>
#9      DioMixin._dispatchRequest
dio_mixin.dart:567
<asynchronous suspension>
```

### PR DESCRIPTION
**Pull request by AlexV525**
Fix #1703.

### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [x] I have updated the `CHANGELOG.md` in the corresponding package

### CODE REVIEW

File: dio/lib/src/dio_mixin.dart [hunk 1]
```diff
@@ -556,18 +556,18 @@ abstract class DioMixin implements Dio {
...
       if (statusOk || reqOpt.receiveDataWhenStatusError == true) {
-        final forceConvert = !(T == dynamic || T == String) &&
-            !(reqOpt.responseType == ResponseType.bytes ||
-                reqOpt.responseType == ResponseType.stream);
-        String? contentType;
-        if (forceConvert) {
-          contentType = headers.value(Headers.contentTypeHeader);
-          headers.set(Headers.contentTypeHeader, Headers.jsonContentType);
+        // Imply as JSON if the given generic type matches the condition.
+        final implicitConvert = T != dynamic &&
```
Comment (ipcjs) [inline dio/lib/src/dio_mixin.dart:560]: `forceConvert` should be disabled when receiving an error status code

```suggestion
        final implicitConvert = statusOk && T != dynamic &&
```

**Review by ipcjs — COMMENTED** [review]
`forceConvert` is used to handle the special case where the generic type of the request method is `Map`/`List` and the server returns `text/plain`, if not forced to convert, a type conversion error will be reported.
The request will throw an exception when it fails, so there is no need to do this processing.

**Comment by AlexV525** [comment]
> `forceConvert` is used to handle the special case where the generic type of the request method is `Map`/`List` and the server returns `text/plain`, if not forced to convert to json, a type conversion error will be reported.

The original issue suggestion is to respect the content type reported from the server. In this case, I would say it's more like normal behavior. WDYT?

**Comment by ipcjs** [comment]
The feature never worked properly. 😂
it's better to remove it（

**Comment by kuhnroyal** [comment]
I think its good to remove it.

**Review by chenglu — APPROVED** [review]
👍

**Pull request merged**